nixos/modules/services/backup/tsm.nix

```diff
@@ -0,0 +1,358 @@
+      # TSM option keys are case insensitive;
+      # we have to ensure there are no keys that
+      # differ only by upper and lower case.
+      type = lib.types.addCheck
+        (lib.types.attrsOf (lib.types.nullOr lib.types.str))
+        (attrs: checkIUnique (lib.attrsets.attrNames attrs));
+      default = {};
+      example.compression = "yes";
+      example.passwordaccess = null;
+      description = ''
+        Additional key-value pairs for the server stanza.
+        Values must be strings, or <literal>null</literal>
+        for the key not to be used in the stanza
+        (e.g. to overrule values generated by other options).
+      '';
+    };
+    options.text = lib.options.mkOption {
+      type = lib.types.lines;
+      example = lib.options.literalExample
+        ''lib.modules.mkAfter "compression no"'';
+      description = ''
+        Additional text lines for the server stanza.
+        This option can be used if certion configuration keys
+        must be used multiple times or ordered in a certain way
+        as the <option>extraConfig</option> option can't
+        control the order of lines in the resulting stanza.
+        Note that the <literal>server</literal>
+        line at the beginning of the stanza is
+        not part of this option's value.
+      '';
+    };
+    options.stanza = lib.options.mkOption {
+      type = lib.types.str;
+      internal = true;
+      visible = false;
+      description = "Server stanza text generated from the options.";
+    };
+    config.name = lib.modules.mkDefault name;
+    # Client system-options file directives are explained here:
+    # https://www.ibm.com/support/knowledgecenter/SSEQVQ_8.1.7/client/c_opt_usingopts.html
+    config.extraConfig =
+      lib.attrsets.mapAttrs (lib.trivial.const lib.modules.mkDefault) (
+        {
+          commmethod = "v6tcpip";  # uses v4 or v6, based on dns lookup result
+          tcpserveraddress = config.server;
+          tcpport = builtins.toString config.port;
+          nodename = config.node;
+          passwordaccess = if config.genPasswd then "generate" else "prompt";
+          passworddir = ''"${config.passwdDir}"'';
+        } // lib.attrsets.optionalAttrs (config.includeExclude!="") {
+          inclexcl = ''"${pkgs.writeText "inclexcl.dsm.sys" config.includeExclude}"'';
+        }
+      );
+    config.text =
+      let
+        attrset = lib.attrsets.filterAttrs (k: v: v!=null) config.extraConfig;
+        mkLine = k: v: k + lib.strings.optionalString (v!="") "  ${v}";
+        lines = lib.attrsets.mapAttrsToList mkLine attrset;
+      in
+        lib.strings.concatStringsSep "\n" lines;
+    config.stanza = ''
+      server  ${config.name}
+      ${config.text}
+    '';
+  };
+
+  options.services.tsmClient = {
+    enable = lib.options.mkEnableOption ''
+      IBM Spectrum Protect (Tivoli Storage Manager, TSM)
+      client system-options file "dsm.sys"
+      used by TSM command line applications
+    '';
```

Comment:
Intertwining `enable` and `servername` this way is unexpected and confusing. The semantics you described warrant, that you split the module in two (code may stay in the same file). One `programs.tsmClient.enable` that draws in the right wrapped binary, and one `services.tsmClient.enable` which enables the service, forcing that `servername` has to be configured and activating `programs.tsmClient.enable`.

Rationale: The way you constructed this module, the tsm-client binaries are expected to work without configuration. Then you should put them into `programs` instead of `services`. The user's expectation is that, when (s)he does `services.*.enable`, the service get's enabled. This silently fails now, if `servername` is not set. I know, that you correctly documented this behavior. Still, this is unexpected and, thus, confusing.